### PR TITLE
[SPARK-47902][SQL]Making Compute Current Time* expressions foldable

### PR DESCRIPTION
**Pull request by dbatomic**
### What changes were proposed in this pull request?

This PR is a followed of [this](https://github.com/apache/spark/pull/44261) PR that made "compute current time" family of expressions `Unevaluable`. Implicitly, all `Unevaluable` expressions are also not foldable. Hence, we have a regression comparing to state prior to 44261 where "compute current time" expressions could have been used in places where constant folding is required.

Proposed change is to keep these expression `Unevaluable` in a sense that `eval`/`codeGen`  can't be called but to allow folding. This is a special case given that these expressions are supposed to be replaced by QO with literals (that are foldable) so proposal is to create new interface for this family of expressions.

### Why are the changes needed?

Explained above.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Additional test is added that ensures that `CurrentDate` can be used in places that require folding.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantFoldingSuite.scala [hunk 1]
```diff
@@ -437,6 +437,21 @@ class ConstantFoldingSuite extends PlanTest {
       Optimize.execute(oneRowScalarSubquery),
       oneRowScalarSubquery)
   }
+
+  test("Current time functions are constant folded") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/ConstantFoldingSuite.scala:441]: I think what we need is an analyzer test to make sure we can use current_time in functions that require folding input, like `rand`. How about we move the test to `ExpressionTypeCheckingSuite`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala [hunk 2]
```diff
@@ -811,4 +811,10 @@ class ExpressionTypeCheckingSuite extends SparkFunSuite with SQLHelper with Quer
         "This should have been converted during analysis."))
     )
   }
+
+  test("check that current time is foldable") {
+    val rnd =
+      Rand(Cast(UnixTimestamp(CurrentDate(), Literal("yyyy-MM-dd HH:mm:ss")), IntegerType))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala:817]: int: we can simply do `Month(CurrentDate())` to turn date into int.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala (hunk 2 above)
Comment (dbatomic) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/ExpressionTypeCheckingSuite.scala]: oh, yes, sure.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**